core: introduce logging API

kaspar030 opened the pull request:
introduce logging API.

(includes log_printnoformat by @LudwigOrtmann)

kaspar030 commented:
I'm thinking of maybe adding a mask to the _log call. That mask could be redefined on a per-module basis as in @authmillenon per-module "ENABLE_DEBUG" so as to have fine-grained log message selection. What do you think?

authmillenon commented on Makefile.dep:
```diff
@@ -1,3 +1,7 @@
+ifneq (,$(filter log_printfnoformat,$(USEMODULE)))
+	DISABLE_MODULE += log_printf
```
Indentation

haukepetersen replied:
still broken :-)

authmillenon commented:
I'd prefer caps for macros (even function-like ones).

authmillenon commented:
Apart from that this looks very good. (But the masks are not added yet, right)

kaspar030 commented:
> I'd prefer caps for macros (even function-like ones).

I agree, but they look so ugly, and the log stuff will be all over the place... Guess we can't make an exception here? ;)

kaspar030 commented:
>  (But the masks are not added yet, right?)

Correct.

authmillenon commented:
>> I'd prefer caps for macros (even function-like ones).
> 
> I agree, but they look so ugly, and the log stuff will be all over the place... Guess we can't make an exception here? ;)

What's speaking against making them inline functions?

kaspar030 commented:
> What's speaking against making them inline functions?

I can't get the vaargs stuff to optimize out.

even if defining ```log``` as
<pre>
static inline void mylog(int level, const char *restrict fmt, ...) {
    if (level <= LOG_LEVEL) {
        va_list args;
        va_start(args, fmt);
        vfprintf(stdout, fmt, args);
        va_end(args);
    }   
}
</pre>

the result is bigger than using the macro printf wrapper. AFAIR @LudwigOrtmann experimented with this, too?

LudwigOrtmann commented:
Regarding va_list: Check the discussion in my PR.

haukepetersen commented on core/include/log.h:
```diff
@@ -0,0 +1,67 @@
+ * @def Log message if level <= LOG_LEVEL and
+ */
+
+#define log(level, ...) if (level <= LOG_LEVEL) _log(level, __VA_ARGS__)
```
I guess its a matter of taste, but I think I would prefer to capitalize the log macros `LOG_ERROR()` etc, to make clear that these trigger these are pre-processor functions...

haukepetersen commented on core/include/log.h:
```diff
@@ -0,0 +1,67 @@
+#include "log_module.h"
+#else
+#include <stdio.h>
+#define _log(level, ...) printf(__VA_ARGS__)
```
I would drop the underscore in favor of something like `log_write` as underscores in RIOT are mostly used for internal stuff...

haukepetersen commented:
+1 for this change.

kaspar030 commented:
- macro names are now upper-case
- documented log_printfnoformat/log_module.h

kaspar030 commented:
alright, this is not WIP anymore.

haukepetersen commented:
By looking at it, I think documenting the log levels rather specifically may help to prevent future miss-usage of those. If we are very clear about when to use e.g. ERROR, than we could maybe save some explaining why (not to) use it in certain situations.

kaspar030 commented:
Totally agree. I tried to give examples on when to use the log levels. What do you think?

OlegHahm commented:
I think I like the solution!

haukepetersen commented:
Looks good. Just one thought: do we need the log-level `LOG_ALL`? It does just exactly the same than `LOG_DEBUG`, doesnt it?! Or should we leave it as convenience log level or something?

kaspar030 commented:
@haukepetersen not sure... It does the same as LOG_DEBUG.  But IMHO "-DLOG_LEVEL=LOG_ALL" might be clearer than "-DLOG_LEVEL=LOG_DEBUG".

haukepetersen commented:
I agree, I was just thinking out loud. So let's leave it in.

kaspar030 commented:
So those are two ACKs? I'll squash then.

kaspar030 commented:
- squashed
- rebased

OlegHahm commented:
> I think I like the solution!

is not an ACK, but I will review ASAP. @LudwigOrtmann, do you have any opinion?

kaspar030 commented on sys/log/log_printfnoformat/log_module.h:
```diff
@@ -0,0 +1,50 @@
+static inline void log_write(unsigned level, const char *format, ...) {
+    (void)level;
+    puts(format);
+}
```
I'm just realizing that this always adds a newline. While a little ugly for normal messages, it completely borks the output when parts of an output line are printed seperately. But replacing with "printf(format)" would keep the printf dependency...

Opinions?

gebart replied:
Can't you use write instead of puts?
On May 7, 2015 2:08 PM, "Kaspar Schleiser" <notifications@github.com> wrote:

> In sys/log/log_printfnoformat/log_module.h
> <https://github.com/RIOT-OS/RIOT/pull/2907#discussion_r29845683>:
>
> > +extern "C" {
> > +#endif
> > +
> > +/**
> > + * @brief log_write overridden function
> > + *
> > + * This example function will only print the log's format string.
> > + * Use it where printf might be too heavy.
> > + *
> > + * @param[in] level (unused)
> > + * @param[in] format String that the function will print
> > + */
> > +static inline void log_write(unsigned level, const char *format, ...) {
> > +    (void)level;
> > +    puts(format);
> > +}
>
> I'm just realizing that this always adds a newline. While a little ugly
> for normal messages, it completely borks the output when parts of an output
> line are printed seperately. But replacing with "printf(format)" would keep
> the printf dependency...
>
> Opinions?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/2907/files#r29845683>.
>

kaspar030 replied:
good idea. but I'm not sure that works on non-newlib platforms.

OlegHahm replied:
Would `puts()` work on these platforms?

gebart replied:
Which platforms are we talking here? I am assuming that this is for the most constrained devices. 
Does avr-libc support printf(), puts(), write()?
What about MSP430?

OlegHahm replied:
Building application "foobar" for "msb-430h" with MCU "msp430f1612".
> What about MSP430?

Apparently not:
```
/home/oleg/git/miniature-dangerzone/playground/playground.c:1:20: fatal error: unistd.h: No such file or directory
...
/home/oleg/git/miniature-dangerzone/playground/bin/msb-430h/foobar.a(playground.o): In function `main'
/home/oleg/git/miniature-dangerzone/playground/playground.c:4: undefined reference to `write'
collect2: ld returned 1 exit status
```

OlegHahm replied:
No solution for this yet, right?

kaspar030 replied:
gcc optimized simple strings without parameters and with newline to puts anyways. It probably does so in this case. OTOH printf would still need to be linkable.

LudwigOrtmann commented:
@OlegHahm I don't have an opinion.

LudwigOrtmann commented:
@kaspar030  Maybe you could also use the logging API, at least in /core.

LudwigOrtmann commented:
Does this API offer the possibility to use different logging modules?
Proper documentation is missing.. Maybe you can copy over and adapt from https://github.com/RIOT-OS/RIOT/pull/2503/files#diff-adf51e1439dc4bb480c2e8b575fe359d

OlegHahm commented:
> Does this API offer the possibility to use different logging modules?

It does, see https://github.com/RIOT-OS/RIOT/pull/2907/files#diff-adf51e1439dc4bb480c2e8b575fe359dR76.

> Proper documentation is missing.. 

I would disagree, but more documentation is always good.

kaspar030 commented:
- added howto for creating log modules
- replaced most core printf/puts calls

OlegHahm commented on core/include/log.h:
```diff
@@ -0,0 +1,99 @@
+ * @addtogroup  core_util
+ * @{
+ *
+ * @file        log.h
```
superfluous, please remove, doxygen can do better without this parameter.

OlegHahm commented on core/include/log.h:
```diff
@@ -0,0 +1,99 @@
+ * @brief       System logging header
+ *
+ * This header offers a bunch of "LOG_*" functions that, with the default
+ * implementation, behave like printf with verbosity levels.
```
`s/behave like/use/`

OlegHahm commented on core/include/log.h:
```diff
@@ -0,0 +1,99 @@
+ *
+ * If desired, it is possible to implement a log module which then will be used
+ * instead the default printf-based implementation.  In order to do so, the log
+ * module has to provide "log_module.h", add -DHAVE_LOG_MODULE to CFLAGS and
```
Hm, I don't like that the module has to modify the CFLAGS somehow. How about defining a PSEUDOMODULE instead?

kaspar030 replied:
actually, that is a good idea. a log module could depend on a pseudo module, which adds the define anyhow. so only one USE_MODULE entry is needed. Will change.

OlegHahm commented on core/include/log.h:
```diff
@@ -0,0 +1,99 @@
+ * The idea is to set LOG_LEVEL to one of these values in the application's Makefile.
+ * That will restrict output of log statements to those with equal or lower log level.
+ *
+ * The default log level is LOG_ALL, which will print every message.
```
Shouldn't we set the default to `LOG_INFO` or `LOG_ERROR` or even `LOG_NONE`?

kaspar030 replied:
No "Hello, this is RIOT!" anymore?

OlegHahm replied:
For a production case I probably rather want to save some kilobytes of memory than being greeted on a UART I will never see.

OlegHahm replied:
But, yes, I think that `LOG_INFO` makes most sense as a default, since >90% of RIOT users are currently RIOT developers and it still can be easily changed.

OlegHahm replied:
I guess the documentation should be updated, too.

LudwigOrtmann commented:
@kaspar030 needs rebase

kaspar030 commented:
- rebased

OlegHahm commented:
What about the default log level?

kaspar030 commented:
sorry, didn't mention last two commits. 
- changed that to INFO, 
- changed the way how to create a logging module, no cflags fiddling required anymore

OlegHahm commented:
Okay, from my side an ACK if you fix the documentation.

kaspar030 commented:
ok, I squashed it, but it needs a second ACK.

kaspar030 commented:
- and rebased again

haukepetersen commented:
ACK

kaspar030 commented:
and GO.
